### DISCUSSION TITLE
Directly in the DB not receive the changes via websockets

### DISCUSSION THREAD

**Discussion by kurdin** [supabase/supabase#298] [Questions] [answered]
Hi guys, thanks for development of interesting project, trying to wrap my head on how is your system works. 
I was playing with next.js example, and everything seems to be working as expecting except this claim from your Docs:

>The beauty of listening to the replication functionality is that you can make changes to your database from anywhere - your API, directly in the DB, via a console etc - and you will still receive the changes via websockets.

After I edited `users` or `todos` table's data directly in DB I don't see any update or change events (in realtime) in the browser. Only if I reload the page I can see updated values from modified db. Am I missing something or misunderstood concept of the system from your Docs?

**Comment by inian** [comment, marked as answer]
Hello Sergey. You can use our [realtime library](https://supabase.io/docs/guides/client-libraries#realtime-changes) to listen to changes. This library will capture realtime changes made from the API, directly via the DB, etc. 

The dashboard in the browser does not update in realtime yet, but we are actively working on this :)

**Reply by kurdin** [reply]
@inian sorry for not providing all information. I did use realtime library and example from it https://github.com/supabase/realtime/tree/master/examples/next-js
but as I said, direct changes to DB does not trigger UI updates / changes via websockets.

**Reply by kiwicopple** [reply]
Hey @kurdin , quick clarifying question - did you set up a project on Supabase, or are you pointing a [Realtime](https://github.com/supabase/realtime) server towards your existing database?

**Reply by kurdin** [reply]
> Hey @kurdin , quick clarifying question - did you set up a project on Supabase, or are you pointing a [Realtime](https://github.com/supabase/realtime) server towards your existing database?

No, I did not use Supabase, I used docker image `supabase/realtime` via docker-compose.yml file in `/realtime/tree/master/examples/next-js` example.

**Reply by kiwicopple** [reply]
Got it! Thanks for clarifying

> After I edited users or todos table's data directly in DB

Looks like we have commented out these lines, so it only works for updates: 

https://github.com/supabase/realtime/blob/b27d71a2d6c48254d1f85aec6122e063dad64a0e/examples/next-js/pages/index.js#L30-L33

If you uncomment them, does it work?

**Reply by kurdin** [reply]
@kiwicopple Yep, it works now with uncommented lines. When do you think this can be ready for production and do you have any mechanism for websockets reconnect if server is down or network is not available ?  Thanks for all your answers.

**Reply by kurdin** [reply]
PS. Seems like WS reconnecting works just fine.

**Reply by kiwicopple** [reply]
> When do you think this can be ready for production

We have a lot of people using it in production already! Including at supabase. I still have to do some benchmarks though. What are you using it for?

> WS reconnecting works just fine

Yes, the server is very fault tolerant (since it's elixir)

**Reply by kurdin** [reply]
@kiwicopple I am working on React Native app with postgraphile as a backend, I want to see if I can use your realtime package to help me sync data across connected users, and see if realtime is more efficient than postgraphile live query.

**Reply by kiwicopple** [reply]
I believe Graphile livequeries also uses WAL: https://www.graphile.org/postgraphile/live-queries/#graphilesubscriptions-lds

It looks like they require wal2json though - Supabase doesn't have that installed. I'm not too sure which would be more efficient just right now sorry @kurdin, but it looks like they hit the database every time a change happens:

> However, we must re-run the user's query to see if anything else (e.g. computed values from functions, views, plugins, etc; reordering of results; effects on pagination) has changed before returning the updated data to the user. Therefore, it’s best to keep live queries small and simple: the more complex the query, the more sources of change it has (so updates may happen more frequently), and the longer it will take to execute.

This probably makes it easier for you, but at the cost to your database

**Reply by kurdin** [reply]
@kiwicopple any plans to make GraphQL module for supabase, similar to postgrest ?

**Reply by kiwicopple** [reply]
No plans, since PostgREST has embedded resources. Is there something you are missing with PostgREST @kurdin ?

**Reply by kurdin** [reply]
@kiwicopple I like REST, simple and strange forward to create and use but GraphQL has a very nice and convenient features that makes it better for larger projects. GraphQL is hot, hype is out there and it's getting more popular among modern devs,  it makes hire smart and progressive devs easy if your project has cool backend in GraphQL. But yeah, some projects just don't need it, REST still a king for regular data communications between UI (App, Web) and Backend (DB).

**Reply by kiwicopple** [reply]
Good feedback, thanks Kurdin - at the moment we want to keep our surface area a small as possible so that we can move fast, but we will definitely consider this for the future

**Reply by oskarer** [reply]
@kurdin If you're looking for GraphQL on Postgres check out [Nhost](https://nhost.io/)
